packages/flutter_tools/lib/src/build_system/targets/dart.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../base/file_system.dart';
+import '../../base/platform.dart';
+import '../../build_info.dart';
+import '../../compile.dart';
+import '../../dart/package_map.dart';
+import '../../project.dart';
+import '../build_system.dart';
+
+/// The define to pass a [BuildMode].
+const String kBuildModeFlag = 'buildMode';
+
+/// Supports compiling dart source to kernel with a subset of flags.
+///
+/// This is a non-incremental compile so the specific [updates] are ignored.
+Future<void> compileKernel(Map<String, ChangeType> updates, Environment environment) async {
+  final KernelCompiler compiler = await kernelCompilerFactory.create(
+    FlutterProject.fromDirectory(environment.projectDir),
+  );
+  final BuildMode buildMode = getBuildModeForName(environment.defines[kBuildModeFlag]);
+  await compiler.compile(
+    aot: buildMode != BuildMode.debug,
+    trackWidgetCreation: false,
+    targetModel: TargetModel.flutter,
+    targetProductVm: buildMode == BuildMode.release,
+    outputFilePath: environment
+      .buildDir
+      .childFile('main.app.dill')
+      .path,
+    depFilePath: null,
+  );
+}
+
+/// Finds the locations of all dart files within the project.
+///
+/// This does not attempt to determine if a file is used or imported, so it
+/// may otherwise report more files than strictly necessary.
+List<SourceFile> listDartSources(Environment environment) {
+  final Map<String, Uri> packageMap = PackageMap(PackageMap.globalPackagesPath).map;
+  final List<SourceFile> dartFiles = <SourceFile>[];
+  for (Uri uri in packageMap.values) {
+    final Directory libDirectory = fs.directory(uri.toFilePath(windows: platform.isWindows));
+    for (FileSystemEntity entity in libDirectory.listSync(recursive: true)) {
+      if (entity is File && entity.path.endsWith('.dart')) {
+        dartFiles.add(SourceFile(entity));
+      }
+    }
+  }
+  return dartFiles;
+}
+
+/// Generate a snapshot of the dart code used in the program.
+const Target kernelSnapshot = Target(
+  name: 'kernel_snapshot',
+  inputs: <Source>[
+    Source.function(listDartSources), // <- every dart file under {PROJECT_DIR}/lib and .packages
```

Comment:
Does the platform.dill need to go here, too?

Reply:
Yes, fixed the cache_dir env to point to bin/cache/artifacts instead of bin/cache/artifacts/engine so we can grab it correctly.

Reply:
err, no its in engine/common - so I didn't need to ... but still seems like a good change

Reply:
I think I don't understand why being in engine/common means you don't need to list it. Put another way, what is the exhaustive list of things that don't need to be listed, and where/how is that documented? Alternately, list *everything* that will be touched by the action. For example, this would additionally list the kernel compiler snapshot.